File: sorting/binary_search/binary_search.py
def binary_search(list, number):
    start, end = 0, len(list) - 1
    while start <= end:
        midpoint = start + (end - start) // 2
        if list[midpoint] == number:
            return midpoint
        else:
            if list[midpoint] <= number:
                start = midpoint + 1

            if list[midpoint] >= number:
                end = midpoint - 1
    
print(binary_search([0, 1, 1, 2, 4, 5, 5, 5, 6, 6, 7, 7, 8, 8, 9], 9))

File: searching/binary_search/binary_search.rs
#![allow(unused_mut)]
fn main() {

    let arr: Vec<i32> = vec![0, 1, 1, 2, 4, 5, 5, 5, 6, 6, 7, 7, 8, 8, 9];
    bin_search(arr, 9)
}
fn bin_search(arr: Vec<i32>, number: i32) {
    let (mut start, mut end) = (0, arr.len());

    while start <= end {
        let mut midpoint = start + (end - start) / 2;

        if arr[midpoint] == number {
            println!("{} found at index {}", number, midpoint);
            break;
        }
        else {
            if arr[midpoint] <= number {
                start = midpoint + 1;
            }
            if arr[midpoint] >= number {
                end = midpoint - 1;
            }
        }
        println!("{} not found", number);
    }
}